src/proto/streams/streams.rs

```diff
@@ -402,44 +404,61 @@ where
         let id = frame.stream_id();
         let promised_id = frame.promised_id();
 
-        let res = {
-            let stream = match me.store.find_mut(&id) {
-                Some(stream) => stream.key(),
-                None => return Err(RecvError::Connection(Reason::PROTOCOL_ERROR)),
-            };
+        // First, ensure that the initiating stream is still in a valid state.
+        let parent_key = match me.store.find_mut(&id) {
+            Some(stream) => {
+                if id > me.actions.recv.max_stream_id() {
+                    trace!("id ({:?}) > max_stream_id ({:?}), ignoring PUSH_PROMISE", id, me.actions.recv.max_stream_id());
+                    return Ok(());
+                }
 
-            if me.counts.peer().is_server() {
-                // The remote is a client and cannot reserve
-                trace!("recv_push_promise; error remote is client");
-                return Err(RecvError::Connection(Reason::PROTOCOL_ERROR));
+                // The stream must be receive open
+                stream.state.ensure_recv_open()?;
+                stream.key()
             }
-
-            me.actions.recv.recv_push_promise(frame,
-                                              &me.actions.send,
-                                              stream,
-                                              &mut me.store)
+            None => return Err(RecvError::Connection(Reason::PROTOCOL_ERROR)),
         };
 
-        if let Err(err) = res {
-            if let Some(ref mut new_stream) = me.store.find_mut(&promised_id) {
+        // Ensure that we can reserve streams
+        me.actions.recv.ensure_can_reserve()?;
 
-                let mut send_buffer = self.send_buffer.inner.lock().unwrap();
-                me.actions.reset_on_recv_stream_err(
-                    &mut *send_buffer,
-                    new_stream,
-                    &mut me.counts,
-                    Err(err))
-            } else {
-                // If there was a stream error, the stream should have been stored
-                // so we can track sending a reset.
-                //
-                // Otherwise, this MUST be an connection error.
-                assert!(!err.is_stream_error());
-                Err(err)
-            }
-        } else {
-            res
+        // Next, open the stream.
+        if me.actions.recv.open(promised_id, true, &mut me.counts)?.is_none() {
+            return Ok(());
```

Comment:
I also had to go look up why `open` might return `None`. Maybe a comment in here about the stream being refused and it is already scheduled?

Maybe we should someday also make the return value of `open` clearer in the first place...